src/cgr_gwas_qc/workflow/modules/bcf.smk

```diff
@@ -0,0 +1,106 @@
+from cgr_gwas_qc import load_config
+from math import ceil
+
+cfg = load_config()
+
+############################
+# General BCF import rules
+###############################
+
+
+def _create_unknown_sex(wildcards):
+    path_to_unknown_sex_lst = temp("unknown_sex.lst")
+    cfg.ss.rename(columns={"Sample_ID": "#IID"}).assign(Sex="U").to_csv(
+        path_to_unknown_sex_lst, sep="\t", index=False
+    )
+    return path_to_unknown_sex_lst
+
+
+rule convert_bcf_to_plink_bed:
+    """Converts BCF to plink BED file
+
+    Path to aggregated BCF file is expected in user_files in config. The expected
+    BCF should be created using BCFtools/gtc2vcf. The BCF will be converted to BED
+    file for processing.
+    """
+    input:
+        bcf=cfg.config.user_files.bcf,
+    params:
+        prefix="sample_level/samples",
+        unknown_sex=_create_unknown_sex,
+    output:
+        bed="sample_level/samples.bed",
+        bim="sample_level/samples.bim",
+        fam="sample_level/samples.fam",
+    log:
+        "sample_level/samples.log",
+    conda:
+        cfg.conda("plink2-0")
+    benchmark:
+        "benchmarks/convert_bcf_to_plink_bed." + str(len(cfg.ss)) + ".tsv"
+    threads: 44
```

Comment:
Thanks for bringing this up. With e91e79270e0c2996ae1f65b72579208cfd6af636, it should now scale down nicely if max. available on system is less than `threads` requested in any rule. I have described it in #351.

The rule still requests the hardcorded 44 to avoid attempts but it would behave better across systems.

Feel free to let me know if you have any additional thoughts/suggestions on this.

Reply:
I received an error for the `convert_bcf_to_plink_bed` rule when I use the generic Slurm profile and specify the max threads to be fewer than 44.  


**command**
` cgr submit --max-threads 24 --slurm`

<br>

**error message**
```
[Thu Nov  7 16:43:52 2024]
rule convert_bcf_to_plink_bed:
    input: sample_level/samples.bcf
    output: sample_level/samples.bed, sample_level/samples.bim, sample_level/samples.fam
    log: sample_level/samples.log                                                                               jobid: 2                                                                                                    benchmark: benchmarks/convert_bcf_to_plink_bed.500.tsv
    threads: 24                                                                                                 resources: tmpdir=/tmp, mem_mb=1059, time_hr=1
                                                                                                            plink2 --allow-extra-chr 0 --keep-allele-order --double-id --bcf sample_level/samples.bcf --update-sex unknown_sex.lst --output-chr 26 --split-par hg38 --make-pgen --out sample_level/bcf2plink  --memory 1059 --threads 24 ;plink2 --pfile sample_level/bcf2plink --make-pgen --sort-vars --out sample_level/bcf2plink-sorted --threads 24 --memory 1059  ;plink2 --pfile sample_level/bcf2plink-sorted --make-bed --out sample_level/samples --threads 24 --memory 1059 ;rm sample_level/bcf2plink.{pgen,psam,pvar,log} sample_level/bcf2plink-sorted.{pgen,psam,pvar,log}
sbatch: error: CPU count per node can not be satisfied
sbatch: error: Batch job submission failed: Requested node configuration is not available
Traceback (most recent call last):
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/github/GwasQcPipeline/src/cgr_gwas_qc/cluster_profiles/slurm_generic/submit.py", line 77, in <module>
    main()
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/github/GwasQcPipeline/src/cgr_gwas_qc/cluster_profiles/slurm_generic/submit.py", line 73, in main
    print(submit_job(cmd))
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/github/GwasQcPipeline/src/cgr_gwas_qc/cluster_profiles/__init__.py", line 144, in submit_job
    return sp.check_output(cmd).decode().rstrip()
  File "/home/marksja/miniconda3/envs/cgr-dev/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/home/marksja/miniconda3/envs/cgr-dev/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['sbatch', '--partition=cgrq', '--job-name=convert_bcf_to_plink_bed.2', '--cpus-per-task=24', '--mem=1gb', '--time=0-1', '--output=logs/convert_bcf_to_plink_bed_2.%j', '--parsable', '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/031/006/.snakemake/tmp.95w3pv14/snakejob.convert_bcf_to_plink_bed.2.sh']' returned non-zero exit status 1.
Error submitting jobscript (exit code 1):

```

Reply:
It would be nice if the `convert_bcf_to_plink_bed` rule was coded something like:
```
threads: $max_threads
```

Reply:
It would be nice if the `convert_bcf_to_plink_bed` rule was coded something like:
```
threads: $max_threads
```

Reply:
It's a bit of usage and possibly documentation issue with the ` cgr submit --max-threads 24 --slurm` error :relaxed:.  Based on the log you shared, the feature is executed perfectly as expected. It is downscaling the threads to 24 as you have requested on the cli with `--max-threads 24`. The issue is you are requesting a 24 cpu node with `slurm_partition: cgrq`. We only have a maximum of 20 cpu node on cgrq partition at CCAD2. One of the following should work fine:

`cgr submit --max-threads <any number less than 20> --slurm`
or with `slurm_partition: defq` in config.yml
`cgr submit --max-threads <any number less than 44> --slurm`

Let me know if the error persists. Thanks.

Reply:
👍